### DISCUSSION TITLE
About the issue of capturing sound card data in Java

### DISCUSSION THREAD

**Discussion by Janix520** [umjammer/vavi-sound#12] [General]
I want to use Java to create a global DSP on Windows, which requires capturing sound card data and filtering the sound before outputting it. Do you have any methods to implement this

**Comment by umjammer** [comment]
see https://www.baeldung.com/java-sound-api-capture-mic

```java

        byte[] buf = new byte[1024];
        int l;
        while (audioStream.available() > 0) {
            l = audioStream.read(buf, 0, buf.length);
            // buf is PCM you can DSP here
            // then output DSP-ed buf to what you want
        }

```

and a DSP library https://github.com/psambit9791/jdsp

**Reply by Janix520** [reply]
This can only capture the microphone, but cannot capture the sound output from the system. I have previously studied that it requires a virtual sound card to create a loopback

**Reply by umjammer** [reply]
it cannot be done by java only.

on macos it needs an application like soundflower, blackhole.
on windows? i'm not sure, see https://groups.google.com/g/jna-users/c/Zt-gK8MKB2Q (keyword is "WASAPI loopback capture"?)
on linux, do `modeprobe snd-aloop`

once you created a loopback device, java would recognize it as one of line in.

**Reply by umjammer** [reply]
see also https://stackoverflow.com/a/27473859

**Comment by Janix520** [comment]
I can successfully capture system sound using the xt-audio library, but after capturing, the original sound is still exist, which means I have taken a copy of the sound and it is meaningless. 

I have searched for many methods, but I will continue to search for other methods. 

Thank you for your answer

**Comment by umjammer** [comment]
note:

macOS has ScreenCaptureKit
